codex/slots/slotbuilder.nim

```diff
@@ -0,0 +1,267 @@
+
+func numRootsPadLeafs*(self: SlotBuilder): Natural =
+  ## Number of padding field elements required for a pow2
+  ## merkle tree for the slot roots.
+  ##
+
+  self.rootsPadLeafs.len
+
+func numSlotBlocks*(self: SlotBuilder): Natural =
+  ## Number of blocks per slot.
+  ##
+
+  self.manifest.blocksCount div self.manifest.numSlots
+
+func numBlockRoots*(self: SlotBuilder): Natural =
+  ## Number of cells per block.
+  ##
+
+  self.manifest.blockSize.int div self.cellSize
+
+func toCellCid(cell: Poseidon2Hash): ?!Cid =
+  let
+    cellMhash = ? MultiHash.init(Pos2Bn128MrklCodec, cell.toBytes).mapFailure
+    cellCid = ? Cid.init(CIDv1, CodexSlotCell, cellMhash).mapFailure
+
+  success cellCid
+
+func toSlotCid(root: Poseidon2Hash): ?!Cid =
+  let
+    mhash = ? MultiHash.init($multiCodec("identity"), root.toBytes).mapFailure
+    treeCid = ? Cid.init(CIDv1, SlotRootCodec, mhash).mapFailure
+
+  success treeCid
+
+func toProvingCid(root: Poseidon2Hash): ?!Cid =
+  let
+    mhash = ? MultiHash.init($multiCodec("identity"), root.toBytes).mapFailure
+    treeCid = ? Cid.init(CIDv1, SlotProvingRootCodec, mhash).mapFailure
+
+  success treeCid
+
+func mapToSlotCids(slotRoots: seq[Poseidon2Hash]): ?!seq[Cid] =
+  success slotRoots.mapIt( ? it.toSlotCid )
+
+func toEncodableProof*(
+  proof: Poseidon2Proof): ?!CodexProof =
+
+  let
+    encodableProof = CodexProof(
+      mcodec: multiCodec("identity"), # copy bytes as is
+      index: proof.index,
+      nleaves: proof.nleaves,
+      path: proof.path.mapIt( @(it.toBytes) ))
+
+  success encodableProof
+
+func toVerifiableProof*(
+  proof: CodexProof): ?!Poseidon2Proof =
+
+  let
+    verifiableProof = Poseidon2Proof(
+      index: proof.index,
+      nleaves: proof.nleaves,
+      path: proof.path.mapIt(
+        ? Poseidon2Hash.fromBytes(it.toArray32).toResult
+      ))
+
+  success verifiableProof
+
+proc getCellHashes*(
+  self: SlotBuilder,
+  slotIndex: int): Future[?!seq[Poseidon2Hash]] {.async.} =
+
+  let
+    treeCid = self.manifest.treeCid
+    blockCount = self.manifest.blocksCount
+    numberOfSlots = self.manifest.numSlots
+
+  logScope:
+    treeCid = treeCid
+    blockCount = blockCount
+    numberOfSlots = numberOfSlots
+    index = blockIndex
+    slotIndex = slotIndex
+
+  let
+    hashes: seq[Poseidon2Hash] = collect(newSeq):
+      for blockIndex in self.strategy.getIndicies(slotIndex):
+        trace "Getting block CID for tree at index"
+
+        without blk =? (await self.store.getBlock(treeCid, blockIndex)), err:
+          error "Failed to get block CID for tree at index"
+          return failure(err)
+
+        without digest =? Poseidon2Tree.digest(blk.data & self.blockPadBytes, self.cellSize), err:
+          error "Failed to create digest for block"
+          return failure(err)
+
+        # TODO: Remove this sleep. It's here to prevent us from locking up the thread.
+        # await sleepAsync(10.millis)
+
+        digest
+
+  success hashes
+
+proc buildSlotTree*(
```

Comment:
Please move the saving of the tree from `buildSlot` to `buildSlotTree`. The data sampler needs to build the tree and save it, *and* have the PoseidonTree object returned to grab proofs.

Reply:
> needs to build the tree and save it,

The sampler doesn't need to save anything, it should build the bottom and top trees of the fly, and use the repostore to access the proofs of the middle tree (which is the large tree and the one that gets persisted).

--------------------------------

There are 3 trees, two are built on the fly, the block trees and the top level slot tree, the middle tree is persisted. 

For the bottom layer, just select the block and use the `digestTree` function from `digest.nim`, and then the `tree.getProof(index)` for the specific block cell index. 

For the middle tree, use the `repostore.getCidAndProof` methods to fetch the required proof. 

The sampler should take a reference to the `RepoStore` and the `SlotBuilder`. The flow should be:

- Get the block where cell is located from the repostore, build a tree on top of it using `diget.digestTree` (or a corresponding `builder.getCellTree` which should take care of the padding), which returns a tree instance
- Invoke `cellTree.getProof(cellIndex)` for the cell proof
- Invoke `repostore.getCidAndProof(cellBlockCid)` to get the proof for the middle persisted tree, and finally, 
- Either invoke a yet to be added `buildSlotsTree` method, or just invoke the `Poseidon2Tree.init(slotRoots & self.rootsPadLeafs)` on it, to get the root of the slots and the corresponding proof on it.

I'll add the `buildCellTree` and `buildSlotsTree` methods as soon as I'm back, which are essentially invocations of `PoseidonTree.init(data & padding)`.

Reply:
This lines up with what I'm trying to do but it's missing one step:
When a slot is first filled, the middle tree doesn't exist in the local blockstore yet. (It was built on the client node, but is not present on the host node.) So, I set the data sampler to fetch the proofs for the middle tree using `repoStore.getCidAndProof`, but when it can't find it, the middle tree has to be created and saved. I figured I would call `slotBuilder.buildSlotTree` to do this, and if it would save the middle tree the problem would be solved. If you'd rather we move the initial middle-tree-building on the host node out of the data sampler, that's fine with me. It just needs to happen somewhere.

Reply:
I would rather not burden the users for `DataSampler` with the need to check the existence of the slot tree and recreate it if necessary. DataSampler is already doing work in this area, so I think the responsibility fits there. I'm working on this, I hope you'll approve: https://github.com/codex-storage/nim-codex/blob/b3da3eb024d2b2a1e0b915daa2a792685e228f49/codex/proof/datasamplerstarter.nim#L72

Reply:
> This lines up with what I'm trying to do but it's missing one step: When a slot is first filled, the middle tree doesn't exist in the local blockstore yet. (It was built on the client node, but is not present on the host node.) So, I set the data sampler to fetch the proofs for the middle tree using `repoStore.getCidAndProof`, but when it can't find it, the middle tree has to be created and saved. I figured I would call `slotBuilder.buildSlotTree` to do this, and if it would save the middle tree the problem would be solved. If you'd rather we move the initial middle-tree-building on the host node out of the data sampler, that's fine with me. It just needs to happen somewhere.

This should really be handled as a descrente step, outside of the sampler. It should happen at the time the node gets assigned the slot. We should simply call `buildSlot` at that moment and by the time the sampler gets called the slot should already be downloaded and stored by the node.